### ISSUE TITLE
Support for partials

### ISSUE THREAD

**Issue by matthewp** [withastro/roadmap#697]
- Accepted Date: 2023-09-13
- Reference Issues/Discussions: https://github.com/withastro/roadmap/discussions/266
- Author: @matthewp 
- Champion(s): @matthewp
- Implementation PR: 

# Summary

Allow pages to be marked as being partials, preventing the doctype tag and any head injection from taking place. A partial is defined as a template which belongs in the context of another page and is dependent on that page in order to be fully functional.

# Background & Motivation

Partials are a technique that has been used by web applications for decades, popularized in frameworks such as Ruby on Rails. Frontend oriented JavaScript frameworks have typically not used partials, but instead use JSON APIs and front-end templating in order to dynamically change parts of the page. Nevertheless, partials have remained a niche feature, and with Astro's backend focus we have had interest in support for a long time.

Recently the popularity of projects like [htmx](https://htmx.org/) and [Unpoly](https://unpoly.com/) have revived interest in this technique. Since Astro treats each page request as a request for a full page it automatically attaches the doctype and head elements, making it difficult to simply inject into the page.

# Goals

- The ability to request a URL from Astro that excludes the usual page elements (doctype, head injection).
- Some way to optionally communicate extra dependencies, such as styles and scripts, so libraries that support head-diffing can achieve that.
- The base feature should work the same in SSG and SSR apps. Partials are still output as `.html` files in a static build.

# Non-Goals 

- This isn't an integration specifically for HTMX or any one library. It should work with any library or manual DOM manipulation that does `innerHTML`.
  - We may take into account special headers from popular libraries if there is a way to enhance the base experience.
- No client-side scripts from Astro will be part of this change.
- Routing should mostly be unchanged; this is metadata about a page but doesn't change the location or where routes live (still in the `pages` folder).

**Comment by lilnasy** [comment]
I would like to bring into the conversation Github.com's web-components-based framework, [catalyst](https://github.com/github/catalyst).

The relevant aspect is that associated javascript and css would still need to be fetched and duplicated.

**Comment by webpro** [comment]
Makes me think of RSS feed. Could this maybe also be used/hooked into when rendering RSS content? So there would be no need for a custom Markdown parser/renderer (e.g. `markdown-it`).

**Comment by ayoayco** [comment]
This is amazing. Would unlock so many use cases for .astro files as API endpoints.

**Comment by aFuzzyBear** [comment]
a bear can only dream....

**Comment by philipschm1tt** [comment]
I am very interested in this in the context of micro-frontends using classic Server-Side Includes.
That is, I would like to include header and footer on a page via SSI as fragments.
Ideally, header, footer, and the body of the page would come from three different Astro applications and be server-side rendered.

**Comment by sasoria** [comment]
Awesome. I'd be very interested in trying this out, with and without htmx!

**Comment by matthewp** [comment]
For htmx users, the [head support](https://htmx.org/extensions/head-support/) extension already does almost exactly what I would want. If you create a page like this:

```html
<head>
  <style>
    .clicked {
      color: blue;
    }
  </style>
</head>
<div class="clicked">I was clicked</div>
```

The styles/scripts are merged into the real head of the document. The fragment target *does not* have any head content.

Is the desire to not use the head support? What is the problem with using Astro with htmx exactly?

**Comment by matthewp** [comment]
Here's my research on head merging support:

Library                       | Head targeting | Head merging
------------------------------|----------------|------------------------
[htmx](https://htmx.org/)     | ✅              | ✅ (head-support plugin)
[Unpoly](https://unpoly.com/) | ✅              | ❌
[jQuery](https://jquery.com/) | ❌              | ❌

**Comment by dougmoscrop** [comment]
I'd like to have a page that might contain multiple components rendered on it and only update one via htmx, without having to rerender the other ones.  Partials/fragments seem like they'd suit this, but there might be a simpler approach I'm not thinking of?

**Comment by michrome** [comment]
> What is the problem with using Astro with htmx exactly?

From the previous discussions, there isn't a _problem_. There's a _desire_ to not have to use [`hx-select`](https://htmx.org/attributes/hx-select/) because the response is a partial not a complete HTML document. This desire is met with the goal:

> * The ability to request a URL from Astro that excludes the usual page elements (doctype, head injection).

---

And what [you're suggesting here](https://github.com/withastro/roadmap/issues/697#issuecomment-1743713007) is perfect for more advanced cases 👍

**Comment by matthewp** [comment]
@michrome Ok thanks. If you use the head-support plugin then the head contents are moved into the head and you don't need to use hx-select. Is that not true? Or do you just not want to use head-support.

**Comment by ZebraFlesh** [comment]
I would like to use Astro partials and am not using htmx.

**Comment by ayoayco** [comment]
🙋‍♂️ Also want to mention I'm thinking of using this for vanilla JS or vanilla custom elements. I think Partials will save developers time and complexity cleaning up HTML or selecting just a portion.

**Comment by michrome** [comment]
First up: what's being proposed above is exactly what I want! 😎

> Ok thanks. If you use the head-support plugin then the head contents are moved into the head and you don't need to use hx-select. Is that not true? Or do you just not want to use head-support.
> 
> I don't consider that to be an advanced case. Astro components contain styles, which is very normal. I would say that not having the styles and dealing with it through global styles is the more advanced case.

The simplest use cases in my mind are:

- appending/replacing `<li>`s in a list; 
- or replacing the text content of a element that's already on screen.

Neither require additional styles. What's being proposed is good because there's no need to use use **`hx-select`** (for htmx users) or do any extraction from the response (for any framework). 👍

Assuming I am being too reductive then perhaps the use case of a new component with styles is more typical. It feels more 'advanced' to me because client side I have to detect <head> content from the rest of the partial … but that's _exactly_ what **`head-support`** is for so I personally wouldn't have any problem using it—in fact I'd _want to use it_ for in this case. 👍

TLDR—this is all good.

**Comment by georgwittberger** [comment]
I have tried the HTMX [head-support](https://htmx.org/extensions/head-support/) extension today and it plays nicely with Astro's current way of injecting styles and scripts into the `<head>` element.

Assuming you have the following initial page markup:

```html
<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="UTF-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>My Page</title>
    <script src="https://unpkg.com/htmx.org@1.9.6"></script>
    <script src="https://unpkg.com/htmx.org@1.9.6/dist/ext/head-support.js"></script>
    <link rel="stylesheet" href="/style.css" />
  </head>
  <body hx-ext="head-support">
    <h1>My Page</h1>
    <button hx-post="/result" hx-swap="outerHTML">Click me!</button>
  </body>
</html>
```

Then the following Astro page `src/pages/result.astro` can be used to swap the `<button>` for new content with potentially new styles:

```html
<html>
  <head></head>
  <body>
    <div class="result">This is pink!</div>
  </body>
</html>

<style>
  .result {
    background-color: hotpink;
  }
</style>
```

Astro injects the `<style>` tag into the `<head>` when sending the response. HTMX then picks it up from there and inserts it into the existing `<head>` element of the page. This happens only once, so even if you would render the `<button>` again inside the result the new `<style>` tag is not inserted over and over again on each request. The `head-support` extension is smart enough to detect which elements have already been inserted.

If you are concerned about having the "real" content inside `html > body` in the Astro page: HTMX also handles that. In this case only the inner HTML of the `<body>` is swapped in. So the resulting HTML document looks like this:

```html
<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="UTF-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>My Page</title>
    <script src="https://unpkg.com/htmx.org@1.9.6"></script>
    <link rel="stylesheet" href="/style.css" />
    <style>.result[data-astro-cid-hhnqfkh6] { background-color: hotpink; }</style>
  </head>
  <body>
    <h1>My Page</h1>
    <div class="result" data-astro-cid-hhnqfkh6>This is pink!</div>
  </body>
</html>
```

**Comment by georgwittberger** [comment]
> @georgwittberger are you inferring partial support isn’t required? If so, I disagree 😀

No, I don't. I just thought it could be helpful to point out what is possible with the current state of Astro.

I agree that a general HTML partial support would be a really useful feature of Astro. However, no matter if you are using HTMX or some other library to patch the DOM client-side the question remains how Astro should support style and script injection for HTML partials. So, if the final decision should be that styles and scripts are simply not supported in HTML partials then I think it could be nice for developers to know how they can achieve some solution with "normal" Astro pages, even if that requires a specific library like HTMX to make head-merging work.

**Comment by matthewp** [comment]
Don't worry everyone, I still intend to add support for partials and am working on the RFC right now! My main concern going into this project was that head merging was something we would need to account for and figure out a solution for. But the fact that htmx has support for it already, and you could build similar things for jQuery and anywhere else, actually means that we *don't* (in my currently opinion) need to try and find a solution for that in Astro. So this can be as simple as a config flag in the page to opt in to partial fragment behavior.

**Comment by georgwittberger** [comment]
@matthewp , this sounds very reasonable. 👍 

I totally agree that partial support in Astro can very well come with limitations, e.g. not being able to inject styles or scripts. The docs can explain such caveats and also workarounds like the one above for those who want to respond with HTML fragments including style / script injection.

**Comment by ayoayco** [comment]
Oh yeah, I don’t expect Astro to handle head merging itself. Will scoped styles be supported still?

**Comment by matthewp** [comment]
@ayoayco No, scoped styles requires head merging. So if you need that you can do that today without any new features just by extracting them from the returned HTML. This proposal will be to explicitly opt-out of styles/scripts.

**Comment by ayoayco** [comment]
Okay clear 👍 partials for plain unstyled HTML. Another question, sorry if out of scope for this story, but in the receiving page of a partial what’s going to be the best approach to style the coming in HTML?

**Comment by matthewp** [comment]
I would assume global CSS, either with `<style is:global>` or imported CSS or an atomic CSS library like Tailwind. I would love for @michrome and others who want this feature to chime in on that.

**Comment by matthewp** [comment]
RFC is up: https://github.com/withastro/roadmap/pull/721 would love feedback there.

**Comment by matthewp** [comment]
This was implemented and is in the stable version of Astro.
